plugins/GranularPitchShifter/GranularPitchShifter.cpp

```diff
@@ -0,0 +1,269 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "GranularPitchShifter.h"
+
+#include <cmath>
+#include "embed.h"
+#include "plugin_export.h"
+
+
+namespace lmms
+{
+
+extern "C"
+{
+Plugin::Descriptor PLUGIN_EXPORT granularpitchshifter_plugin_descriptor =
+{
+	LMMS_STRINGIFY(PLUGIN_NAME),
+	"Granular Pitch Shifter",
+	QT_TRANSLATE_NOOP("PluginBrowser", "Granular pitch shifter"),
+	"Lost Robot <r94231/at/gmail/dot/com>",
+	0x0100,
+	Plugin::Type::Effect,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	nullptr,
+} ;
+}
+
+
+GranularPitchShifterEffect::GranularPitchShifterEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key) :
+	Effect(&granularpitchshifter_plugin_descriptor, parent, key),
+	m_granularpitchshifterControls(this),
+	m_prefilter({PrefilterLowpass(), PrefilterLowpass()})
+{
+	autoQuitModel()->setValue(autoQuitModel()->maxValue());
+	
+	changeSampleRate();
+}
+
+
+bool GranularPitchShifterEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning()) { return false; }
+
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * pitchBuf = m_granularpitchshifterControls.m_pitchModel.valueBuffer();
+	const ValueBuffer * pitchSpreadBuf = m_granularpitchshifterControls.m_pitchSpreadModel.valueBuffer();
+
+	const float size = m_granularpitchshifterControls.m_sizeModel.value();
+	const float shape = m_granularpitchshifterControls.m_shapeModel.value();
+	const float jitter = m_granularpitchshifterControls.m_jitterModel.value();
+	const float twitch = m_granularpitchshifterControls.m_twitchModel.value();
+	const float spray = m_granularpitchshifterControls.m_sprayModel.value();
+	const float spraySpread = m_granularpitchshifterControls.m_spraySpreadModel.value();
+	const float density = m_granularpitchshifterControls.m_densityModel.value();
+	const float glide = m_granularpitchshifterControls.m_glideModel.value();
+	const int minLatency = m_granularpitchshifterControls.m_minLatencyModel.value() * m_sampleRate;
+	const float densityInvRoot = std::sqrt(1.f / density);
+	const float feedback = m_granularpitchshifterControls.m_feedbackModel.value() * densityInvRoot;
+	const float fadeLength = 1.f / m_granularpitchshifterControls.m_fadeLengthModel.value();
+	const bool prefilter = m_granularpitchshifterControls.m_prefilterModel.value();
+	
+	if (glide != m_oldGlide)
+	{
+		m_oldGlide = glide;
+		m_glideCoef = std::exp(-1 / (glide * m_sampleRate));
+	}
+	
+	const float shapeK = cosWindowApproxK(shape);
+	const int sizeSamples = static_cast<int>((m_sampleRate / size) * 0.5) * 2;
```

Comment:
Do you really need to divide by 0.5, and multiply by 2? Wouldn't this cancel out?

Reply:
No, it wouldn't cancel out because of the integer cast.  This forces sizeSamples to be an even number, which used to be required, but no longer is since waitMult was changed to a float.  I'll fix this.